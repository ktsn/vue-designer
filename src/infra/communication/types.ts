export interface WebSocketServer {
  clients: Set<WebSocket>
  on(event: 'connection', cb: (socket: WebSocket) => void): void
}

export interface WebSocket {
  send(payload: string): void
  on(event: 'message', cb: (message: string) => void): void
  on(event: 'error', cb: (error: Error) => void): void
  once(event: 'close', cb: () => void): void
}

export type Resolver = Record<string, (...args: any[]) => any>

export type Mutator = Record<string, (...args: any[]) => void>
