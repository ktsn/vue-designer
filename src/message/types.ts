import { DeclarationForUpdate, DeclarationForAdd } from '../parser/style/types'
import { VueFilePayload } from '../parser/vue-file'

export interface Events {
  initClient: undefined
  selectNode: {
    uri: string
    templatePath: number[]
    stylePaths: number[][]
  }
  addNode: {
    currentUri: string
    nodeUri: string
    path: number[]
  }
  changeActiveEditor: string
  updateEditor: {
    uri: string
    code: string
  }
  addDocument: {
    uri: string
    code: string
  }
  changeDocument: {
    uri: string
    code: string
  }
  removeDocument: string
  addDeclaration: {
    uri: string
    path: number[]
    declaration: DeclarationForAdd
  }
  removeDeclaration: {
    uri: string
    path: number[]
  }
  updateDeclaration: {
    uri: string
    declaration: DeclarationForUpdate
  }
}

export interface Commands {
  initProject: Record<string, VueFilePayload>
  changeDocument: string
  highlightEditor: {
    uri: string
    ranges: [number, number][]
  }
  updateEditor: {
    uri: string
    code: string
  }
}
