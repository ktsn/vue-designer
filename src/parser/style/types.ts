import { Range } from '../modifier'

export interface HasChildren<T extends ChildNode> {
  children: T[]
}

export interface Style extends Range, HasChildren<AtRule | Rule> {
  path: [number]
}

export interface Rule extends Range, HasChildren<Declaration> {
  type: 'Rule'
  before: string
  after: string
  path: number[]
  selectors: Selector[]
}

export interface Declaration extends Range {
  type: 'Declaration'
  before: string
  after: string
  path: number[]
  prop: string
  value: string
  important: boolean
}

export interface AtRule extends Range, HasChildren<ChildNode> {
  type: 'AtRule'
  before: string
  after: string
  path: number[]
  name: string
  params: string
}

export type ChildNode = AtRule | Rule | Declaration

export interface Selector {
  type: 'Selector'
  universal: boolean
  tag?: string
  id?: string
  class: string[]
  attributes: Attribute[]
  pseudoClass: PseudoClass[]
  pseudoElement?: PseudoElement
  leftCombinator?: Combinator
}

export interface PseudoClass {
  type: 'PseudoClass'
  value: string
  params: Selector[]
}

export interface PseudoElement {
  type: 'PseudoElement'
  value: string
  pseudoClass: PseudoClass[]
}

export type AttributeOperator = '=' | '~=' | '|=' | '^=' | '$=' | '*='

export interface Attribute {
  type: 'Attribute'
  operator?: AttributeOperator
  name: string
  value?: string
  insensitive: boolean
}

export interface Combinator {
  type: 'Combinator'
  operator: string
  left: Selector
}

/*
 * Used to print in the preview
 */
export interface RuleForPrint {
  path: number[]
  selectors: string[]
  children: DeclarationForPrint[]
}

export interface DeclarationForPrint {
  path: number[]
  prop: string
  value: string
}

/**
 * Used to update ast
 */
export interface DeclarationForAdd {
  prop: string
  value: string
  important: boolean
}

export interface DeclarationForUpdate {
  path: number[]
  prop?: string
  value?: string
  important?: boolean
}
