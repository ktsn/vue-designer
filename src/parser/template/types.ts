import { Range } from '../modifier'

export type ElementChild = Element | TextNode | ExpressionNode

interface BaseNode extends Range {
  type: string
}

export interface Template extends BaseNode {
  type: 'Template'
  attrs: Record<string, Attribute>
  children: ElementChild[]
}

export interface Element extends BaseNode {
  type: 'Element'
  path: number[]
  name: string
  startTag: StartTag
  endTag?: EndTag
  children: ElementChild[]
}

export interface StartTag extends BaseNode {
  type: 'StartTag'

  // It simply maps attributes with the following rule:
  // 1. Static attributes are mapped to `attrs`
  // 2. Normal v-bind attributes are mapped to `props`
  // 3. v-bind attributes having `prop` modifier are mapped to `domProps`
  // 4. Other directives are in `directives`
  // `attrs` and `props` may include `class` and `style` unlike Vue.js template compiler.
  // Object style v-bind (v-bind="obj") is not supported yet. It will be in `directives`.
  attrs: Record<string, Attribute>
  props: Record<string, Directive>
  domProps: Record<string, Directive>
  directives: Directive[]

  selfClosing: boolean
}

export interface EndTag extends BaseNode {
  type: 'EndTag'
}

export interface TextNode extends BaseNode {
  type: 'TextNode'
  path: number[]
  text: string
}

export interface ExpressionNode extends BaseNode {
  type: 'ExpressionNode'
  path: number[]
  expression: string
}

export interface Attribute extends BaseNode {
  type: 'Attribute'
  attrIndex: number
  name: string
  value?: string
}

export interface Directive extends BaseNode {
  type: 'Directive'
  attrIndex: number
  name: string
  argument?: string
  modifiers: Record<string, boolean>
  expression?: string
  value?: any
}

export interface VForDirective extends Directive {
  name: 'for'
  left: string[]
  right?: string
}
