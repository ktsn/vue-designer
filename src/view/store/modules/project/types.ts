import { Template } from '@/parser/template/types'
import { Prop, Data, ChildComponent } from '@/parser/script/types'

export interface ScopedDocument {
  uri: string
  displayName: string
  template: Template | undefined
  props: Prop[]
  data: Data[]
  childComponents: ChildComponent[]
  styleCode: string
}

export interface DocumentScopeItem {
  type: string | null
  value: any
}

export interface DocumentScope {
  props: Record<string, DocumentScopeItem>
  data: Record<string, DocumentScopeItem>
}

export type DraggingPlace = 'before' | 'after' | 'first' | 'last'
