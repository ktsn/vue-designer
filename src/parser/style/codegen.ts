import assert from 'assert'
import * as t from './types'

export function genStyle(ast: t.STStyle): string {
  return ast.children
    .map(node => {
      switch (node.type) {
        case 'AtRule':
          return genAtRule(node)
        case 'Rule':
          return genRule(node)
        default:
          return assert.fail(
            `[style codegen] Unexpected node type ${(node as any).type} on root`
          )
      }
    })
    .join('\n')
}

function genAtRule(atRule: t.STAtRule): string {
  let buf = `@${atRule.name} ${atRule.params}`

  if (atRule.children.length > 0) {
    const children = atRule.children
      .map(child => {
        switch (child.type) {
          case 'AtRule':
            return genAtRule(child)
          case 'Rule':
            return genRule(child)
          case 'Declaration':
            return genDeclaration(child)
          default:
            return assert.fail(
              `[style codegen] Unexpected node type ${
                (child as any).type
              } as child of AtRule`
            )
        }
      })
      .join(' ')

    buf += ' {' + children + '}'
  } else {
    buf += ';'
  }

  return buf
}

export function genRule(rule: t.STRule): string {
  const selectors = rule.selectors.map(genSelector).join(', ')
  const declarations = rule.children.map(genDeclaration).join(' ')

  return `${selectors} {${declarations}}`
}

export function genSelector(s: t.STSelector): string {
  let buf = ''
  if (s.universal) {
    buf += '*'
  } else if (s.tag) {
    buf += s.tag
  }

  if (s.id) {
    buf += '#' + s.id
  }

  if (s.class.length > 0) {
    buf += s.class.map(c => '.' + c).join('')
  }

  if (s.attributes.length > 0) {
    const attrsCodes = s.attributes.map(attr => {
      if (attr.operator != null && attr.value != null) {
        const suffix = attr.insensitive ? ' i' : ''
        const quote = attr.value.includes('"') ? "'" : '"'
        return `[${attr.name}${attr.operator}${quote}${
          attr.value
        }${quote}${suffix}]`
      } else {
        return `[${attr.name}]`
      }
    })
    buf += attrsCodes.join('')
  }

  if (s.pseudoClass.length > 0) {
    buf += s.pseudoClass.map(genPseudoClass).join('')
  }

  if (s.pseudoElement) {
    buf += '::' + s.pseudoElement.value

    const pseudoClass = s.pseudoElement.pseudoClass
    if (pseudoClass.length > 0) {
      buf += pseudoClass.map(genPseudoClass).join('')
    }
  }

  if (s.leftCombinator) {
    buf =
      genSelector(s.leftCombinator.left) +
      ' ' +
      s.leftCombinator.operator +
      ' ' +
      buf
  }

  return buf
}

function genPseudoClass(node: t.STPseudoClass): string {
  if (node.params.length > 0) {
    const selectors = node.params.map(genSelector).join(', ')
    return ':' + node.value + '(' + selectors + ')'
  } else {
    return ':' + node.value
  }
}

export function genDeclaration(decl: t.STDeclaration): string {
  let buf = decl.prop + ': ' + decl.value

  if (decl.important) {
    buf += ' !important'
  }

  return buf + ';'
}
