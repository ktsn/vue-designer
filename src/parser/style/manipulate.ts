import * as t from './types'
import assert from 'assert'
import { clone, unquote } from '../../utils'
import { AssetResolver } from '../../asset-resolver'

interface StyleVisitor {
  atRule?(atRule: t.AtRule): t.AtRule | void
  rule?(rule: t.Rule): t.Rule | void
  declaration?(decl: t.Declaration): t.Declaration | void
  lastSelector?(selector: t.Selector, rule: t.Rule): t.Selector | void
}

export const scopePrefix = 'data-scope-'

function visitStyle(style: t.Style, visitor: StyleVisitor): t.Style {
  function apply<T>(node: T, visitor?: (node: T) => T | void): T {
    return visitor ? visitor(node) || node : node
  }

  function loop(
    node: t.AtRule | t.Rule | t.Declaration
  ): t.AtRule | t.Rule | t.Declaration {
    switch (node.type) {
      case 'AtRule':
        return clone(apply(node, visitor.atRule), {
          children: node.children.map(loop)
        })
      case 'Rule':
        return clone(apply(node, visitor.rule), {
          selectors: node.selectors.map(selector => {
            return visitor.lastSelector && isInterestSelector(node, style)
              ? visitor.lastSelector(selector, node) || selector
              : selector
          }),
          children: node.children.map(loop)
        })
      case 'Declaration':
        return apply(node, visitor.declaration)
    }
  }

  return clone(style, {
    children: style.children.map(loop)
  })
}

export function visitLastSelectors(
  root: t.Style,
  fn: (selector: t.Selector, rule: t.Rule) => t.Selector | void
): t.Style {
  return visitStyle(root, { lastSelector: fn })
}

/**
 * Excludes selectors in @keyframes
 */
function isInterestSelector(rule: t.Rule, root: t.Style): boolean {
  const atRules = rule.path
    .slice(1)
    .reduce<t.ChildNode[]>((nodes, index) => {
      const prev = (nodes[nodes.length - 1] || root) as t.HasChildren<
        t.ChildNode
      >
      assert(
        'children' in prev,
        '[style manipulate] the rule probably has an invalid path.'
      )

      return nodes.concat(prev.children[index])
    }, [])
    .filter((n): n is t.AtRule => n.type === 'AtRule')

  return atRules.every(r => !/-?keyframes$/.test(r.name))
}

export function resolveAsset(
  style: t.Style,
  basePath: string,
  resolver: AssetResolver
): t.Style {
  return visitStyle(style, {
    declaration: decl => {
      const value = decl.value
      const replaced = value.replace(/url\(([^)]+)\)/g, (_, p) => {
        const unquoted = unquote(p)
        const resolved = resolver.pathToUrl(unquoted, basePath)
        return 'url(' + (resolved ? JSON.stringify(resolved) : p) + ')'
      })
      return replaced !== value
        ? clone(decl, {
            value: replaced
          })
        : decl
    }
  })
}

export function addScope(node: t.Style, scope: string): t.Style {
  const keyframes = new Map<string, string>()

  const keyframesReplaced = visitStyle(node, {
    atRule(atRule) {
      if (/-?keyframes$/.test(atRule.name)) {
        const replaced = atRule.params + '-' + scope
        keyframes.set(atRule.params, replaced)

        return clone(atRule, {
          params: replaced
        })
      }
    }
  })

  return visitStyle(keyframesReplaced, {
    declaration: decl => {
      // individual animation-name declaration
      if (/^(-\w+-)?animation-name$/.test(decl.prop)) {
        return clone(decl, {
          value: decl.value
            .split(',')
            .map(v => keyframes.get(v.trim()) || v.trim())
            .join(',')
        })
      }

      // shorthand
      if (/^(-\w+-)?animation$/.test(decl.prop)) {
        return clone(decl, {
          value: decl.value
            .split(',')
            .map(v => {
              const vals = v.trim().split(/\s+/)
              const i = vals.findIndex(val => keyframes.has(val))
              if (i !== -1) {
                vals.splice(i, 1, keyframes.get(vals[i])!)
                return vals.join(' ')
              } else {
                return v
              }
            })
            .join(',')
        })
      }
    },

    lastSelector: selector => {
      return clone(selector, {
        attributes: selector.attributes.concat({
          type: 'Attribute',
          name: scopePrefix + scope,
          insensitive: false
        })
      })
    }
  })
}

export function getNode(
  styles: t.Style[],
  path: number[]
): t.ChildNode | undefined {
  return path.reduce<any>(
    (acc, i) => {
      if (!acc) return

      if (acc.children) {
        return acc.children[i]
      }
    },
    { children: styles }
  )
}

export function getDeclaration(
  styles: t.Style[],
  path: number[]
): t.Declaration | undefined {
  const res = getNode(styles, path)
  return res && res.type === 'Declaration' ? res : undefined
}
