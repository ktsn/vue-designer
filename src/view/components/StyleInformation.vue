<template>
  <ul class="style-information">
    <li
      v-for="(rule, ruleIndex) in rules"
      :key="rule.path.join('.')"
      class="rule"
      @click="onClickRule(rule, ruleIndex)"
    >
      <p
        class="selector-list"
        @click.stop="onClickSelectorList(rule, ruleIndex)"
      >
        <span
          v-for="s in rule.selectors"
          :key="s"
          class="selector"
        >{{ s }}</span>
      </p>

      <ul class="declaration-list">
        <li
          v-for="(decl, declIndex) in rule.children"
          :key="decl.path.join('.')"
          class="declaration"
          @click.stop="onClickDeclaration(decl, ruleIndex, declIndex)"
        >
          <StyleDeclaration
            :prop="decl.prop"
            :value="decl.value"
            :auto-focus="shouldFocusFor(ruleIndex, declIndex)"
            @update:prop="updateDeclarationProp(decl.path, arguments[0])"
            @update:value="updateDeclarationValue(decl.path, arguments[0])"
            @remove="removeDeclaration(decl.path)"
            @input-start:prop="onStartStyleInput"
            @input-start:value="onStartStyleInput"
            @input-end:prop="onEndStyleInput(ruleIndex, declIndex, 'prop', arguments[0])"
            @input-end:value="onEndStyleInput(ruleIndex, declIndex, 'value', arguments[0])"
          />
        </li>
      </ul>
    </li>
  </ul>
</template>

<script lang="ts">
import Vue from 'vue'
import StyleDeclaration from './StyleDeclaration.vue'
import { STRuleForPrint, STDeclarationForPrint } from '@/parser/style/types'

export default Vue.extend({
  name: 'StyleInformation',

  components: {
    StyleDeclaration
  },

  props: {
    rules: {
      type: Array as () => STRuleForPrint[],
      required: true
    }
  },

  data() {
    return {
      autoFocusTarget: undefined as
        | { rule: number; declaration: number; type: 'prop' | 'value' }
        | undefined,
      endingInput: false
    }
  },

  created() {
    const vm: any = this
    vm.endingInputTimer = null
  },

  methods: {
    updateDeclarationProp(path: number[], prop: string): void {
      this.$emit('update-declaration', {
        path,
        prop
      })
    },

    updateDeclarationValue(path: number[], value: string): void {
      this.$emit('update-declaration', {
        path,
        value
      })
    },

    addDeclaration(
      rulePath: number[],
      ruleIndex: number,
      declIndex: number
    ): void {
      this.$emit('add-declaration', {
        path: rulePath.concat(declIndex)
      })

      // Focus on the new declaration prop after it is added
      const unwatch = this.$watch('rules', () => {
        this.autoFocusTarget = {
          rule: ruleIndex,
          declaration: declIndex,
          type: 'prop'
        }
        unwatch()
      })
    },

    removeDeclaration(path: number[]): void {
      this.$emit('remove-declaration', { path })
    },

    onClickRule(rule: STRuleForPrint, index: number): void {
      if (this.endingInput) return

      this.addDeclaration(rule.path, index, rule.children.length)
    },

    onClickSelectorList(rule: STRuleForPrint, index: number): void {
      if (this.endingInput) return

      this.addDeclaration(rule.path, index, 0)
    },

    onClickDeclaration(
      decl: STDeclarationForPrint,
      ruleIndex: number,
      declIndex: number
    ): void {
      if (this.endingInput) return

      const rulePath = decl.path.slice(0, -1)
      this.addDeclaration(rulePath, ruleIndex, declIndex + 1)
    },

    /*
     * While it will add a new rule when clicking empty space,
     * we don't want to do that if it is intended to cancel style editing.
     * To resolve that situation, we limit to add a new declaration
     * when the user is starting editing style value. Then release
     * the limitation after several time passes from finished editing.
     */
    onStartStyleInput(): void {
      const vm: any = this
      clearTimeout(vm.endingInputTimer)

      this.endingInput = true
      this.autoFocusTarget = undefined
    },

    onEndStyleInput(
      rule: number,
      decl: number,
      type: string,
      meta: { reason: string }
    ): void {
      // If the user end the input by pressing enter or tab key,
      // focus on the next form.
      if (meta.reason === 'enter' || meta.reason === 'tab') {
        this.focusOnNextForm(rule, decl, type)
      }

      const delayToEndEdit = 200
      const vm: any = this

      clearTimeout(vm.endingInputTimer)
      vm.endingInputTimer = setTimeout(() => {
        this.endingInput = false
      }, delayToEndEdit)
    },

    focusOnNextForm(rule: number, decl: number, type: string): void {
      if (type === 'prop') {
        this.autoFocusTarget = {
          rule,
          declaration: decl,
          type: 'value'
        }
        return
      }

      if (type === 'value') {
        this.autoFocusTarget = {
          rule,
          declaration: decl + 1,
          type: 'prop'
        }

        // If it is the last item, add a new rule
        const targetRule = this.rules[rule]
        if (decl + 1 === targetRule.children.length) {
          this.$emit('add-declaration', {
            path: targetRule.path.concat(targetRule.children.length)
          })
        }
      }
    },

    shouldFocusFor(rule: number, decl: number): 'prop' | 'value' | null {
      const f = this.autoFocusTarget
      return f && f.rule === rule && f.declaration === decl ? f.type : null
    }
  }
})
</script>

<style scoped>
.style-information,
.declaration-list {
  margin: 0;
  padding: 0;
  list-style: none;
}

.rule:not(:first-child) {
  margin-top: 13px;
  padding-top: 10px;
  border-top: 1px solid var(--vd-border-color);
}

.rule::after {
  content: '}';
}

.selector-list {
  margin: 0;
}

.selector-list::after {
  content: ' {';
}

.selector:not(:first-child)::before {
  content: ', ';
}

.declaration-list {
  padding-left: 1em;
}
</style>
