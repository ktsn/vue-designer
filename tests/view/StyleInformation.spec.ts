import { shallowMount, Wrapper } from '@vue/test-utils'
import StyleInformation from '@/view/components/StyleInformation.vue'
import { STRuleForPrint, STDeclarationForPrint } from '@/parser/style/types'

describe('StyleInformation', () => {
  const StyleDeclaration = {
    name: 'StyleDeclaration',
    props: ['prop', 'value', 'autoFocus'],
    render(this: any, h: Function) {
      return h('div', {
        attrs: {
          styleDeclarationStub: true,
          prop: this.prop,
          value: this.value,
          autoFocus: this.autoFocus
        }
      })
    }
  }

  const rules: STRuleForPrint[] = [
    {
      path: [0],
      selectors: ['a'],
      children: [
        {
          path: [0, 0],
          prop: 'color',
          value: 'red'
        },
        {
          path: [0, 1],
          prop: 'font-size',
          value: '22px'
        }
      ]
    }
  ]

  const create = () => {
    return shallowMount(StyleInformation, {
      propsData: {
        rules
      },
      stubs: {
        StyleDeclaration
      }
    })
  }

  const toDeclarationHtml = (wrapper: Wrapper<any>) => {
    return wrapper
      .findAll('[styledeclarationstub]')
      .wrappers.map(w => w.html())
      .join('\n')
  }

  describe('moving focus', () => {
    it('does not move focus if editing is ended by blur', () => {
      const wrapper = create()
      wrapper
        .findAll(StyleDeclaration)
        .at(0)
        .vm.$emit('input-end:prop', { reason: 'blur' })

      expect(toDeclarationHtml(wrapper)).toMatchSnapshot()
    })

    it('moves from prop to value', () => {
      const wrapper = create()
      wrapper
        .findAll(StyleDeclaration)
        .at(0)
        .vm.$emit('input-end:prop', { reason: 'enter' })

      expect(toDeclarationHtml(wrapper)).toMatchSnapshot()
    })

    it('moves from value to prop', () => {
      const wrapper = create()
      wrapper
        .findAll(StyleDeclaration)
        .at(0)
        .vm.$emit('input-end:value', { reason: 'tab', shiftKey: true })

      expect(toDeclarationHtml(wrapper)).toMatchSnapshot()
    })

    it('moves from value to next prop value', () => {
      const wrapper = create()
      wrapper
        .findAll(StyleDeclaration)
        .at(0)
        .vm.$emit('input-end:value', { reason: 'enter' })

      expect(toDeclarationHtml(wrapper)).toMatchSnapshot()
    })

    it('moves from prop to prev declaration value', () => {
      const wrapper = create()
      wrapper
        .findAll(StyleDeclaration)
        .at(1)
        .vm.$emit('input-end:prop', { reason: 'tab', shiftKey: true })

      expect(toDeclarationHtml(wrapper)).toMatchSnapshot()
    })

    it('removes focus if it cannot go back', () => {
      const wrapper = create()
      wrapper
        .findAll(StyleDeclaration)
        .at(0)
        .vm.$emit('input-end:prop', { reason: 'tab', shiftKey: true })

      expect(toDeclarationHtml(wrapper)).toMatchSnapshot()
    })

    it('adds a new declaration and moves focus to it', () => {
      const wrapper = create()
      wrapper
        .findAll(StyleDeclaration)
        .at(1)
        .vm.$emit('input-end:value', { reason: 'enter' })

      expect(wrapper.emitted('add-declaration')[0][0]).toEqual({
        path: [0, 2]
      })

      wrapper.setProps({
        rules: [
          {
            ...rules[0],
            children: [
              ...rules[0].children,
              {
                path: [0, 2],
                prop: 'property',
                value: 'value'
              }
            ]
          }
        ]
      })

      expect(toDeclarationHtml(wrapper)).toMatchSnapshot()
    })
  })

  describe('add a new declaration', () => {
    const newDecl = {
      path: [0, 99],
      prop: 'font-weight',
      value: 'bold'
    }

    const addTo = (decl: STDeclarationForPrint, index: number) => {
      return [
        {
          ...rules[0],
          children: [
            ...rules[0].children.slice(0, index),
            decl,
            ...rules[0].children.slice(index)
          ]
        }
      ]
    }

    it('adds to the first position', async () => {
      const wrapper = create()
      wrapper.find('.selector-list').trigger('click')

      expect(wrapper.emitted('add-declaration')[0][0]).toEqual({
        path: [0, 0]
      })

      wrapper.setProps({
        rules: addTo(newDecl, 0)
      })

      await wrapper.vm.$nextTick()

      expect(toDeclarationHtml(wrapper)).toMatchSnapshot()
    })

    it('adds to the last position', async () => {
      const wrapper = create()
      wrapper.find('.rule').trigger('click')

      expect(wrapper.emitted('add-declaration')[0][0]).toEqual({
        path: [0, 2]
      })

      wrapper.setProps({
        rules: addTo(newDecl, 2)
      })

      await wrapper.vm.$nextTick()

      expect(toDeclarationHtml(wrapper)).toMatchSnapshot()
    })

    it('inserts to the clicked position', async () => {
      const wrapper = create()
      wrapper.find('.declaration').trigger('click')

      expect(wrapper.emitted('add-declaration')[0][0]).toEqual({
        path: [0, 1]
      })

      wrapper.setProps({
        rules: addTo(newDecl, 1)
      })

      await wrapper.vm.$nextTick()

      expect(toDeclarationHtml(wrapper)).toMatchSnapshot()
    })
  })
})
