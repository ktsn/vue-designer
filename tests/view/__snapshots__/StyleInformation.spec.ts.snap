// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleInformation add a new declaration adds to the first position 1`] = `
<div styledeclarationstub="true" prop="font-weight" value="bold" autofocus="prop"></div>
<div styledeclarationstub="true" prop="color" value="red"></div>
<div styledeclarationstub="true" prop="font-size" value="22px"></div>
`;

exports[`StyleInformation add a new declaration adds to the last position 1`] = `
<div styledeclarationstub="true" prop="color" value="red"></div>
<div styledeclarationstub="true" prop="font-size" value="22px"></div>
<div styledeclarationstub="true" prop="font-weight" value="bold" autofocus="prop"></div>
`;

exports[`StyleInformation add a new declaration inserts to the clicked position 1`] = `
<div styledeclarationstub="true" prop="color" value="red"></div>
<div styledeclarationstub="true" prop="font-weight" value="bold" autofocus="prop"></div>
<div styledeclarationstub="true" prop="font-size" value="22px"></div>
`;

exports[`StyleInformation moving focus adds a new declaration and moves focus to it 1`] = `
<div styledeclarationstub="true" prop="color" value="red"></div>
<div styledeclarationstub="true" prop="font-size" value="22px"></div>
<div styledeclarationstub="true" prop="property" value="value" autofocus="prop"></div>
`;

exports[`StyleInformation moving focus does not move focus if editing is ended by blur 1`] = `
<div styledeclarationstub="true" prop="color" value="red"></div>
<div styledeclarationstub="true" prop="font-size" value="22px"></div>
`;

exports[`StyleInformation moving focus moves from prop to prev declaration value 1`] = `
<div styledeclarationstub="true" prop="color" value="red" autofocus="value"></div>
<div styledeclarationstub="true" prop="font-size" value="22px"></div>
`;

exports[`StyleInformation moving focus moves from prop to value 1`] = `
<div styledeclarationstub="true" prop="color" value="red" autofocus="value"></div>
<div styledeclarationstub="true" prop="font-size" value="22px"></div>
`;

exports[`StyleInformation moving focus moves from value to next prop value 1`] = `
<div styledeclarationstub="true" prop="color" value="red"></div>
<div styledeclarationstub="true" prop="font-size" value="22px" autofocus="prop"></div>
`;

exports[`StyleInformation moving focus moves from value to prop 1`] = `
<div styledeclarationstub="true" prop="color" value="red" autofocus="prop"></div>
<div styledeclarationstub="true" prop="font-size" value="22px"></div>
`;

exports[`StyleInformation moving focus removes focus if it cannot go back 1`] = `
<div styledeclarationstub="true" prop="color" value="red"></div>
<div styledeclarationstub="true" prop="font-size" value="22px"></div>
`;
